File: src/main/kotlin/dev/jefy/connectpro/management/domain/Badge.java
package dev.jefy.connectpro.management.domain;


import org.springframework.util.Assert;

import dev.jefy.connectpro.management.domain.vo.BadgeId;
import dev.jefy.connectpro.management.domain.vo.HexColor;
import dev.jefy.connectpro.shared.infrastructure.ddd.DAggregateRoot;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

/**
 * @author Jôph Yamba
 */
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
@Table(name = "badges")
public class Badge implements DAggregateRoot<BadgeId> {
    @EmbeddedId
    @AttributeOverride(name = "value", column = @Column(name = "value"))
    private BadgeId id;
    
    private String name;
    
    @Embedded
    @AttributeOverride(name = "value", column = @Column(name = "color"))
    private HexColor color; 
    
    private String description; 

    public Badge(String name, HexColor color, String description) {
        validate(name, color, description);
        this.id = BadgeId.generate();
        this.name = name;
        this.color = color;
        this.description = description;
    }
    
    public void update(String name, HexColor color, String description) {
        validate(name, color, description);
        this.name = name;
        this.color = color; 
        this.description = description;
    }
    
    private void validate(String name, HexColor color, String description){
        Assert.notNull(name, "name cannot be null");
        Assert.notNull(color, "value cannot be null");
        Assert.notNull(description, "description cannot be null");
    }
}

/*
    Verified
    Top Rated
    Pro Seller
    Trusted
    New Seller
 */

File: src/main/kotlin/dev/jefy/connectpro/management/domain/Badge.kt
package dev.jefy.connectpro.management.domain

import dev.jefy.connectpro.management.domain.vo.BadgeId
import dev.jefy.connectpro.management.domain.vo.HexColor
import jakarta.persistence.*

@Entity
@Table(name = "badges")
open class Badge(name: String, color: HexColor, description: String) {

    @EmbeddedId
    @AttributeOverride(name = "value", column = Column(name = "value"))
    var id: BadgeId = BadgeId.generate()
        protected set

    var name: String = name
        protected set

    @Embedded
    @AttributeOverride(name = "value", column = Column(name = "color"))
     var color: HexColor = color
        protected set

    var description: String = description
        protected set

    init {
        require(name.isNotBlank()) { "Badge name must not be blank" }
        require(description.isNotBlank()) { "Award color must not be blank" }
    }

    fun update(name: String, color: HexColor, description: String) {
        require(name.isNotBlank()) { "Badge name must not be blank" }
        require(description.isNotBlank()) { "Award color must not be blank" }
        this.name = name
        this.color = color
        this.description = description
    }
    
}